keras/preprocessing/sequence.py

```diff
@@ -273,85 +272,200 @@ class TimeseriesGenerator(Sequence):
+    assert np.array_equal(x[1,0], np.array([10]))
+
+    print "** test 3"
+    data_gen = TimeseriesGenerator(data, targets, length=10, sampling_rate=3, batch_size=2)
+
+    assert len(data_gen) == 10
+
+    print "** test 4 (stateful)"
+    data_gen = TimeseriesGenerator(data, targets, length=10, sampling_rate=2, batch_size=12, stateful=True)
+
+    assert data_gen.stride == 2
+    assert data_gen.batch_size == 10
+
+    print "** test 5 (text sequences seq2one)"
+    txt = bytearray("Keras is simple.")
+    data_gen = TimeseriesGenerator(txt, txt, length=10, batch_size=1)
+
+    for i in range(len(data_gen)):
+        print data_gen[i][0].tostring(), " -> '%s'" % data_gen[i][1].tostring()
+
+    assert data_gen[-1][0].shape == (1, 10) and data_gen[-1][1].shape==(1,)
+    assert data_gen[-1][0].tostring() == u" is simple"
+    assert data_gen[-1][1].tostring() == u"."
+
+    print "** test 6 (text sequences seq2seq)"
+    data_gen = TimeseriesGenerator(txt, txt, length=10, target_seq=True)
+
+    assert data_gen[-1][0].shape == (1, 10) and data_gen[-1][1].shape==(1,10,1)
+    for i in range(len(data_gen)):
+        print data_gen[i][0].tostring(), " -> '%s'" % data_gen[i][1].tostring()
+
+    assert data_gen[0][1].tostring() == u"eras is si"
+    
 
-    batch_0 = data_gen[0]
-    x, y = batch_0
-    assert np.array_equal(x,
-                          np.array([[[0], [2], [4], [6], [8]],
-                                    [[1], [3], [5], [7], [9]]]))
-    assert np.array_equal(y,
-                          np.array([[10], [11]]))
     ```
     """
 
     def __init__(self, data, targets, length,
                  sampling_rate=1,
                  stride=1,
-                 start_index=0,
-                 end_index=None,
-                 shuffle=False,
-                 reverse=False,
-                 batch_size=128):
-        self.data = data
-        self.targets = targets
+                 gap = None,
+                 start_index = 0,
+                 end_index = None,
+                 shuffle = False,
+                 reverse = False,
+                 target_seq = False,
+                 batch_size = 1,
+                 dtype = None,
+                 stateful = False):
+        
+        assert length > 0
+        assert sampling_rate > 0
+        assert batch_size > 0
+        assert len(data) <= len(targets)
+        
+        self.data = np.asarray(data)
+        self.targets = np.asarray(targets)
+        
+        # FIXME: seems required by sparse losses on seq output
+        if target_seq and len(self.targets.shape)<2:
+            self.targets = np.expand_dims(targets,axis=-1)
+        
+        if dtype is None:
+            self.data_type = self.data.dtype
+            self.targets_type = self.targets.dtype
+        else:
+            self.data_type = dtype
+            self.targets_type = dtype
+            
+        
+        # force stateful-compatible parameters
+        if stateful:
+            shuffle=False
```

Comment:
I don't think you should silently change the parameters like this, because this generates unexpected and potentially unwanted behavior in the user's perspective.  
For example, suppose I have a certain amount of GRAM that allows me to train with 256 samples per batch. If I set `batch_size=256`, I expect training to run at full capacity. This code, however, might drastically reduce the amount of samples in each batch, which will not fill all memory. Because the operation is silent, it might take me a very long time to figure it out why I'm running at suboptimal capacity.

For that reason, I think it would be best to just raise an error stating that these parameters are not compatible and which changes should be made to correct the incompatibility.

Reply:
Ok, I will change the stateful option : if set, it will only check if the other parameters are 'stateful-compatible' and rise a ValueError otherwise.